rotary_utils/repair.py

```diff
@@ -622,105 +494,147 @@ def run_end_repair(long_read_filepath: str, assembly_fasta_filepath: str, assemb
+         pd.DataFrame({'contig': contig_info.linear_contig_names, 'circular': 'N', 'repaired': 'N'})], axis=0)
+    repair_info.to_csv(repair_paths.end_repair_status_filepath, sep='\t', index=False)
+
+
+def repair_handle_failed_contigs(assembly_info, repair_paths, failed_contig_names, keep_failed_contigs, output_dir):
+    """
+    Repair and handle failed contigs in the assembly.
+
+    :param assembly_info: An object containing information about the assembly.
+    :param repair_paths: An object containing repair paths.
+    :param failed_contig_names: A list of names of contigs that could not be circularized.
+    :param keep_failed_contigs: A boolean indicating whether to keep the original (non-repaired) versions of
+                                failed contigs in the final output file.
+    :param output_dir: The directory to which the output files should be saved.
+    """
+    if os.path.isdir(os.path.join(repair_paths.linking_outdir_base, 'troubleshooting')):
+        shutil.move(os.path.join(repair_paths.linking_outdir_base, 'troubleshooting'),
+                    os.path.join(output_dir, 'troubleshooting'))
+    if keep_failed_contigs is False:
+        logger.error(f'{len(failed_contig_names)} contigs could not be circularized. A partial output file '
+                     f'including successfully circularized contigs (and no linear contigs) is available at '
+                     f'{repair_paths.end_repaired_contigs_filepath} for debugging. Exiting with error status. See temporary '
+                     f'files and verbose logs for more details.')
+        sys.exit(1)
+    elif keep_failed_contigs is True:
+        logger.warning(f'{len(failed_contig_names)} contigs could not be circularized. The original (non-repaired) '
+                       f'versions of these contigs will be included in the final output file')
+        logger.warning(f'Names of contigs that could not be circularized: {", ".join(failed_contig_names)}')
+
+        # Get the non-repaired circular contigs and append them to the repaired contigs file
+        with open(repair_paths.end_repaired_contigs_filepath, 'a') as append_handle:
+            for record in subset_sequences(assembly_info.assembly_fasta_filepath, failed_contig_names):
+                SeqIO.write(record, append_handle, 'fasta')
+    else:
+        error = ValueError(f'keep_failed_contigs should be a boolean True or False; you provided '
+                           f'{keep_failed_contigs}')
+        logger.error(error)
+        raise error
+
+
+def repair_no_circular_input_contigs(assembly, repair_paths):
+    """
+    Handle repair when no circular contigs are found.
+    """
+    logger.info('No circular contigs. Will copy the input file, repaired_info.tsv, and finish early.')
+    shutil.copyfile(assembly.assembly_fasta_filepath, repair_paths.end_repaired_contigs_filepath)
+    linear_contig_names = parse_assembly_info_file(assembly.assembly_info_filepath, assembly.assembly_info_type,
```

Comment:
Could this move moved out to the parent function?

Reply:
Given that `run_end_repair` is already quite long, I am OK with keeping this as a separate function for now, even though the function is quite short.